uaclient/config.py

```diff
@@ -665,6 +700,37 @@ def help(self, name):
         response_dict["help"] = help_ent.help_info
         return response_dict
 
+    def write_cfg(self, config_path=None):
+        """Write config values back to config_path or DEFAULT_CONFIG_FILE."""
+        if not config_path:
+            config_path = DEFAULT_CONFIG_FILE
+        content = "# Ubuntu-Advantage client config file.\n"
+
+        for attr in ("contract_url", "security_url"):
+            content += "{key}: '{value}'\n".format(
+                key=attr, value=getattr(self, attr)
+            )
+        content += "data_dir: {}\n".format(self.data_dir)
+        if "log_level" in self.cfg:
+            log_level = self.cfg["log_level"].lower()
+        else:
+            log_level = CONFIG_DEFAULTS["log_level"]
+        content += "log_level: {}\n".format(log_level)
+        content += "log_file: {}\n".format(self.log_file)
```

Comment:
Can we use `yaml.dump` for everything instead of hand-crafting this section?

Reply:
I agree that we could easily run `yaml.dump` here, but this will make some things a little bit harder:

1) We will loose the comments we are making in the file. Even though the header comment is easy to add with the `yaml.dump` solution, comments in the middle may not be as easy. But maybe we don't need those comments in the middle.
2) This code is removing any invalid configuration from `uaclient.conf`, like user configurations with typos. However, since we expect users to not directly interact with the file and those incorrect configs will not affect uaclient, maybe this is not needed.
3) This code is keeping the order of the keys. Maybe we can sort the keys in an order that we want, before using it on `yaml.dump`.

Those are not strong arguments to not use `yaml.dump`, just some points. If we can address them, I think we should use `yaml.dump`.

Am I missing something here @blackboxsw ?

Reply:
Nothing missed other than the intent leave some minor documentation breadcrumb comments in /etc/ubuntu-advantage/uaclient.conf so people editing that file directly are alerted to the fact that there are better places (more synchronous places) to make those changes. Filed and issue to track this outstanding discussion point to make sure we have a public-facing decision on what we ultimately decide to to here.
https://github.com/canonical/ubuntu-advantage-client/issues/1702